packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -148,64 +155,87 @@ export function toVisualTree(tree: VisualSharedTreeNode): VisualChildNode {
+ * Concatenate allowed types for `ObjectNodeStoredSchema` and `MapNodeStoredSchema`.
  */
-function concatenateTypes(fieldTypes: ReadonlySet<string>): string {
+export function concatenateTypes(fieldTypes: ReadonlySet<string>): string {
 	return [...fieldTypes].join(" | ");
 }
 
-/**
- * Returns the allowed fields & types for the object fields (e.g., `foo : string | number, bar: boolean`)
- */
-function getObjectAllowedTypes(schema: SimpleObjectNodeSchema): string {
-	const result: string[] = [];
-
-	for (const [fieldKey, treeFieldSimpleSchema] of Object.entries(schema.fields)) {
-		const fieldTypes = treeFieldSimpleSchema.allowedTypes;
-		result.push(`${fieldKey} : ${concatenateTypes(fieldTypes)}`);
-	}
-
-	return `{ ${result.join(", ")} }`;
-}
-
 /**
  * Returns the schema & fields of the node.
  */
 async function visualizeVerboseNodeFields(
-	tree: VerboseTreeNode,
-	treeSchema: SimpleTreeSchema,
+	treeFields: VerboseTree[] | Record<string, VerboseTree>,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	allowedTypes: Record<string, ReadonlySet<string>>,
+	requirements: Record<string, boolean> | undefined,
 	visualizeChildData: VisualizeChildData,
 ): Promise<Record<string, VisualSharedTreeNode>> {
-	const treeFields = tree.fields;
-
 	const fields: Record<string | number, VisualSharedTreeNode> = {};
 
 	for (const [fieldKey, childField] of Object.entries(treeFields)) {
-		fields[fieldKey] = await visualizeSharedTreeNodeBySchema(
+		fields[fieldKey] = await visualizeSharedTreeBySchema(
 			childField,
-			treeSchema,
+			treeDefinitions,
+			allowedTypes[fieldKey],
+			requirements === undefined ? requirements : requirements[fieldKey],
 			visualizeChildData,
 		);
 	}
 
 	return fields;
 }
 
+/**
+ * Extracts and stores allowed types & kind from {@link SimpleFieldSchema}.
+ */
+function storeObjectAllowedTypes(schema: SimpleObjectNodeSchema): {
+	allowedTypes: Record<string, ReadonlySet<string>>;
+	elementIsRequired: Record<string, boolean>;
+} {
+	const allowedTypes: Record<string, ReadonlySet<string>> = {};
+
+	/**
+	 * Maps each field or node identifier to a boolean indicating
+	 * whether it is required (true) or optional (false).
+	 */
+	const elementIsRequired: Record<string, boolean> = {};
+
+	for (const [fieldKey, treeFieldSimpleSchema] of Object.entries(schema.fields)) {
+		allowedTypes[fieldKey] = treeFieldSimpleSchema.allowedTypes;
+		elementIsRequired[fieldKey] =
+			treeFieldSimpleSchema.kind === FieldKind.Required ? true : false;
+	}
+
+	return { allowedTypes, elementIsRequired };
+}
+
 /**
  * Returns the schema & fields of the node with type {@link ObjectNodeStoredSchema}.
  */
 async function visualizeObjectNode(
 	tree: VerboseTreeNode,
-	nodeSchema: SimpleObjectNodeSchema,
-	treeSchema: SimpleTreeSchema,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	allowedTypes: ReadonlySet<string>,
+	isRequired: boolean | undefined,
```

Comment:
I'm assuming that the `undefined` value for `isRequired` means "don't display requirement properties in the tooltip? We should be sure to document that, if so.

Reply:
Also, we bundle `allowedTypes` and `isRequired` together in a few places. It might make sense to extract those into an interface ("FieldSchemaProperties" or something). That would give you a singular place to document the `undefined` semantics of `isRequired`).